wikiclass/feature_lists/wikidatawiki.py

```diff
@@ -21,6 +39,137 @@ class items:
+				for i, source in enumerate(list_of_claims.sources):
+					for index_list_1, index_list_2 in source.items(): 
+						sources_in_JSON = index_list_2[0].toJSON()
+						list_sources_in_JSON.append(sources_in_JSON['datavalue']['value'])
+		
+		return list_sources_in_JSON
+	except:
+		return 0 #if the revision does not contain any sources, return 0
+
+def _process_wikimedia_sources(item):
+
+	list_sources_in_JSON = []
+	list_wikimedia_sources = []
+	
+	for property in item.claims:
+		list_of_properties = property
+		for claim in item.claims[list_of_properties]:
+			list_of_claims = claim
+			for i, source in enumerate(list_of_claims.sources):
+				for index_list_1, index_list_2 in source.items(): 
+					sources_in_JSON = index_list_2[0].toJSON()
+					list_sources_in_JSON.append(sources_in_JSON['datavalue']['value'])
+    
+	wdir_path = os.path.dirname(os.path.realpath(__file__)) #get the current working directory path
+	csv_path = os.path.join(wdir_path, 'excluded_qids.csv')
+
+	for list_sources_in_JSON_content in list_sources_in_JSON: 
+		with open(csv_path) as csvfile:
+			readCSV = csv.reader(csvfile, delimiter=',')
+			for line in readCSV:				
+				try:
+					if(list_sources_in_JSON_content['numeric-id'] == int(line[0])): #if a source comes from Wikimedia projects and DBpedia (i.e. DBpedia collects data from Wikipedia), append that source to the list
+						list_wikimedia_sources.append(list_sources_in_JSON_content)
+						break
+				except:
+					continue
+	return list_wikimedia_sources
+	
+def _process_unique_sources(item):
+	
+	list_sources_in_JSON = []
+	result_set = set()
+	
+	for property in item.claims:
+		list_of_properties = property
+		for claim in item.claims[list_of_properties]:
+			list_of_claims = claim
+			for i, source in enumerate(list_of_claims.sources):
+				for index_list_1, index_list_2 in source.items(): 
+					sources_in_JSON = index_list_2[0].toJSON()
+					list_sources_in_JSON.append("property : " + str(sources_in_JSON['property']) + " & value : " + str(sources_in_JSON['datavalue']['value']))
+	
+	#eliminate duplicates
+	for value in list_sources_in_JSON:
+		result_set.add(value)
+	
+	return len(result_set)
+	
+
+def _process_complete_translations(item_labels, item_descriptions):
+	item_labels_dict = item_labels
+	item_desc_dict = item_descriptions
+	combined_dict = {}
+	result_set = []
+	
+	#merging item label dictionary and item description dictionary
+	for key in (item_labels_dict.keys() | item_desc_dict.keys()):
+		if key in item_labels_dict: combined_dict.setdefault(key, []).append(item_labels_dict[key])
+		if key in item_desc_dict: combined_dict.setdefault(key, []).append(item_desc_dict[key])
+	
+	#if a language consists of both item label and item description exist, add the language into result_set
+	for value in combined_dict.items():
+		if(len(value[1]) == 2): 
+			result_set.append(value[0])
+			
+	return len(result_set)
+
+def _process_important_translations(item_labels, item_descriptions):
```

Comment:
What are we trying to evaluate in such feature? 

I thought that if an item has partially complete important translations, it could be measured by this _process_important_translations functions. 
For example, there are two items - an item which have 4 completed important translations and an item which have 8 completed important translations (i.e. this item has all important translations). The function will give a score 0.5 for the former, and 1 for the latter. Using the score, we can see the former has less completed important translations than the latter.